src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -259,6 +259,142 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
             }
         }
 
+        /// <inheritdoc />
+        public override async Task CopyToAsync(PipeWriter destination, CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            CancellationTokenSource tokenSource = InternalTokenSource;
+            if (tokenSource.IsCancellationRequested)
+            {
+                ThrowHelper.ThrowOperationCanceledException_ReadCanceled();
+            }
+
+            CancellationTokenRegistration reg = default;
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.UnsafeRegister(state => ((StreamPipeReader)state!).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                try
+                {
+                    if (_bufferedBytes > 0)
+                    {
+                        BufferSegment? segment = _readHead;
+                        try
+                        {
+                            while (segment != null)
+                            {
+                                FlushResult flushResult = await destination.WriteAsync(segment.Memory, tokenSource.Token).ConfigureAwait(false);
+
+                                if (flushResult.IsCanceled)
+                                {
+                                    ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
+                                }
+
+                                segment = segment.NextSegment;
+
+                                if (flushResult.IsCompleted)
+                                {
+                                    return;
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            // Advance even if WriteAsync throws so the PipeReader is not left in the
+                            // currently reading state
+                            if (segment != null)
```

Comment:
How would segment be null if `_bufferedBytes > 0`? Is code just being defensive?

Reply:
I have removed the `_bufferebBytes > 0` check as it not required anymore.